src/anyio/_core/_contextmanagers.py

```diff
@@ -102,10 +112,12 @@ class AsyncContextManagerMixin(Generic[_T_co]):
         that once you enter it, you can't re-enter before first exiting it.
     """
 
-    __cm: AbstractAsyncContextManager[_T_co] | None = None
+    __cm: AbstractAsyncContextManager[object] | None = None
 
     @final
-    async def __aenter__(self) -> _T_co:
+    async def __aenter__(self: _SupportsAsyncCtxMgr[_T_co]) -> _T_co:
+        # Needed for mypy to assume self still has the __cm member
+        assert isinstance(self, AsyncContextManagerMixin)
```

Comment:
pyright treats the self typing as an implicit intersection with the containing class but mypy doesn't (even if it does other magic inference fine).